contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,364 @@
+
+#### The `VolumeSnapshot` Object
+
+```GO
+
+// The volume snapshot object accessible to the user. Upon successful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotContent through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents the desired state of the snapshot
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+        // PersistentVolumeClaimName is the name of the PVC being snapshotted
+        // +optional
+        PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+        // SnapshotContentName binds the VolumeSnapshot object with the VolumeSnapshotContent
+        // +optional
+        SnapshotContentName string `json:"snapshotContentName" protobuf:"bytes,2,opt,name=snapshotContentName"`
+
+        // Name of the VolumeSnapshotClass required by the volume snapshot.
+        // +optional
+        VolumeSnapshotClassName string `json:"storageClassName" protobuf:"bytes,3,opt,name=storageClassName"`
+}
+
+type VolumeSnapshotStatus struct {
+        // CreatedAt is the time when the snapshot was successfully created. When it is set,
+	// users can start to resume/unfreeze the application if it was previously frozen 
+	// before taking the snapshot.
+        // +optional
+        CreatedAt *metav1.Time `json:"createdAt" protobuf:"bytes,1,opt,name=createdAt"`
+
+        // AvailableAt is the time when the snapshot was successfully created and available
+	// to use. Some cloud volume providers upload the snapshot to cloud storage after
+	// it is created, so the snapshot will become available some time after snapshot is created.
+	// For other plugins, the snapshot might be available at the same time when the snapshot 
+	// is created.
+        // +optional
+        AvailableAt *metav1.Time `json:"availableAt" protobuf:"bytes,2,opt,name=availableAt"`
+	
+	// Setting Bound to true means that the snapshot is created (and uploaded for cloud providers) sucuessfully, and 
+	// VolumeSnapshotContent is created and bound with the VolumeSnapshot. After Bound becomes true, users
+	// can start to use VolumeSnapshot to provision new volumes.
+	Bound bool `json:"bound" protobuf:"varint,4,opt,name=bound"`
```

Comment:
Can we rename this to ready? Arguably the snapshot is bound when it is created it is just not read to be use.

bound to me implies that something is associated which is true once the snapshot is created. ready seem more appropriate as the volume is ready to use.